feat(routes): use named routes in core

hypeJunction opened the pull request:
Rewrites page handlers using new named route API.
Updates some resource views with logic contained in removed page handler functions.
Deprecates the use of handler parameter in entity menus and removes it
for some core entities.
Add elgg_generate_entity_url().
Add elgg_generate_action_url().

Refs #11563
Fixes #9126

mrclay commented on mod/activity/elgg-plugin.php:
```diff
@@ -19,4 +19,13 @@
 	],
+	'routes' => [
+		'activity' => [
+			'path' => '/activity/{page_type}/{username?}',
```
I think better to have multiple routes. Even though same view can handle them, different logical resources.

hypeJunction replied:
Sure

hypeJunction commented:
@mrclay Thoughts on `elgg_generate_entity_url()`? It's a bit corky and will probably fail at some point, but it is doing the job until proven otherwise.

hypeJunction committed:
feat(routes): use named routes in core

Rewrites page handlers using new named route API.
Updates some resource views with logic contained in removed page handler functions.
Deprecates the use of handler parameter in entity menus and removes it
for some core entities.
Add elgg_generate_entity_url().
Add elgg_generate_action_url().

Refs #11563
Fixes #9126

mrclay commented on engine/lib/pagehandler.php:
```diff
@@ -58,6 +58,101 @@ function elgg_generate_url($name, array $parameters = []) {
+	};
+
+	$pairs = [
+		[$entity->getType(), $entity->getSubtype()],
```
`->type`, `->subtype` thanks to you.

mrclay commented on engine/lib/pagehandler.php:
```diff
@@ -58,6 +58,101 @@ function elgg_generate_url($name, array $parameters = []) {
+				break;
+
+			default :
+				$query[$prop] = $entity->$prop;
```
Use case for this?

hypeJunction replied:
`guid`, `container_guid`, `owner_guid` etc.
I imagine someone using `time_created` or some other metadata in the route

mrclay commented on engine/lib/pagehandler.php:
```diff
@@ -58,6 +58,101 @@ function elgg_generate_url($name, array $parameters = []) {
+ * @param ElggEntity $entity      Entity
+ * @param string     $resource    Resource name
+ * @param string     $subresource Subresource name
+ * @param array      $query       URL query elements
```
Rather this be $params. $query makes me think querystring.

hypeJunction replied:
Anything that's not defined as a URL segment goes into query string

hypeJunction replied:
I will rename and update the docs for both

mrclay commented on engine/lib/pagehandler.php:
```diff
@@ -58,6 +58,101 @@ function elgg_generate_url($name, array $parameters = []) {
+ */
+function elgg_generate_action_url($action, array $query = [], $add_csrf_tokens = true) {
+
+	$url = "action/$action";
```
Yes! Never wondering if it's "actions/" or "action/".

hypeJunction replied:
Not to mention `"action/friends/add?guid=" . elgg_get_page_owner_guid()`

jdalsem commented on engine/lib/navigation.php:
```diff
@@ -627,39 +627,47 @@ function _elgg_page_menu_setup(\Elgg\Hook $hook) {
+			$delete_url = elgg_generate_action_url($action, [
+				'guid' => $entity->guid,
+			]);
+		}
```
should fallback to `entity/delete`

jdalsem commented:
needs rebase

hypeJunction commented:
FFS
